index.js

```diff
@@ -11,96 +11,185 @@ const Type = (exports.Type = {
 
-    if ((mergeGeometry || autoGenerateShape) && !sceneRoot) {
-      console.warn("cannot use mergeGeometry or autoGenerateShape if sceneRoot is null");
+    if (options.offset) {
+      offset.set(options.offset.x, options.offset.y, options.offset.z);
+    } else {
+      offset.set(0, 0, 0);
+    }
+
+    if (options.orientation) {
+      orientation.set(options.orientation.x, options.orientation.y, options.orientation.z, options.orientation.w);
+    } else {
+      orientation.set(0, 0, 0, 1);
+    }
+
+    if (fit !== FIT.MANUAL && !sceneRoot) {
+      console.warn("cannot use all or compound fit if sceneRoot is null");
       return null;
     }
 
-    const computeRadius = function() {
-      _computeBounds(sceneRoot, mergeGeometry, bounds);
-      _computeSphere(sceneRoot, mergeGeometry, bounds, sphere);
+    const computeRadius = function(root) {
+      _computeBounds(root, fit, bounds);
+      _computeSphere(root, fit, bounds, sphere);
       return sphere.radius;
     };
 
-    const computeHalfExtents = function() {
-      _computeBounds(sceneRoot, mergeGeometry, bounds);
-      return halfExtents
+    const computeHalfExtents = function(root) {
+      _computeBounds(root, fit, bounds);
+      halfExtents
         .subVectors(bounds.max, bounds.min)
         .multiplyScalar(0.5)
         .clampScalar(minHalfExtent, maxHalfExtent);
+      localOffset
+        .addVectors(bounds.max, bounds.min)
+        .multiplyScalar(0.5)
+        .applyMatrix4(matrix);
+      return halfExtents;
     };
 
-    //TODO: Support convex hull decomposition, compound shapes, gimpact (dynamic trimesh)
-    let collisionShape;
-    switch (type) {
-      case Type.BOX: {
-        const hx = autoGenerateShape ? computeHalfExtents() : options.halfExtents;
-        collisionShape = _createBoxShape(hx);
-        break;
-      }
-      case Type.CYLINDER: {
-        const hx = autoGenerateShape ? computeHalfExtents() : options.halfExtents;
-        collisionShape = _createCylinderShape(hx, cylinderAxis);
-        break;
-      }
-      case Type.CAPSULE: {
-        const hx = autoGenerateShape ? computeHalfExtents() : options.halfExtents;
-        collisionShape = _createCapsuleShape(hx, cylinderAxis);
-        break;
-      }
-      case Type.CONE: {
-        const hx = autoGenerateShape ? computeHalfExtents() : options.halfExtents;
-        collisionShape = _createConeShape(hx, cylinderAxis);
-        break;
-      }
-      case Type.SPHERE: {
-        const radius = !isNaN(options.sphereRadius) ? options.sphereRadius : computeRadius();
-        collisionShape = new Ammo.btSphereShape(radius);
-        break;
-      }
-      case Type.HULL: {
-        collisionShape = _createHullShape(sceneRoot, mergeGeometry, margin, options.hullMaxVertices || 100000);
-        break;
-      }
-      case Type.MESH: {
-        collisionShape = _createTriMeshShape(sceneRoot, mergeGeometry);
-        break;
+    const createCollisionShape = function(root) {
```

Comment:
For me, it would be time to break these into separate top-level functions, I am starting to get confused about what variables are being used for which parts.